cts/cli/regression.acls.exp

```diff
@@ -372,14 +372,9 @@ crm_attribute: Error performing operation: Permission denied
 =#=#=#= End test: unknownguy: Set stonith-enabled - Insufficient privileges (4) =#=#=#=
 * Passed: crm_attribute  - unknownguy: Set stonith-enabled
 =#=#=#= Begin test: unknownguy: Create a resource =#=#=#=
-pcmk__check_acl 	trace: User 'unknownguy' without ACLs denied read/write access to /cib/configuration/resources/primitive[@id]
-pcmk__apply_creation_acl 	trace: Creation of <primitive> scaffolding with id="<unset>" is implicitly allowed
+pcmk__check_acl 	trace: User 'unknownguy' without ACLs denied read/write access to /cib/configuration/resources/primitive[@id='dummy']
+pcmk__apply_creation_acl 	trace: ACLs disallow creation of <primitive> with id="dummy"
```

Comment:
Thank you, I missed this in the sea of epoch changes.

I don't know yet but it's definitely worth looking into.

Reply:
From previous commit message:
> The only meaningful difference between this [update_cib_object()] and
> pcmk__xml_copy() is that this checks ACLs after copying a set of
> attributes and returns an error upon ACL denial. That error appears in
> the failed result XML from cib_process_create().
> 
> However, pcmk__xml_copy() still handles change tracking for ACL
> purposes. If there's an ACL denial, it will be caught later by
> cib_perform_op(), and the update will be discarded.
> 
> Catching a denial later is a solid tradeoff for the reduction in
> complexity IMO.

Previously, ACLs denied creation of attributes in the `<primitive>` element, and that caused an immediate error in `update_cib_object()`. Now, the error is caught only when ACLs are checked for the change as a whole later.

So we don't get a `<failed>` object because there's no longer a failure within `cib_process_create() -> add_cib_object()` like there was before.

Side note, `cibadmin` doesn't support `--output-as=xml` yet. This `<failed>` object only appears as undocumented, schema-less XML in text output.

> do you know why it happened?

been trying to figure out EXACTLY what differences exist...

Reply:
> So we don't get a `<failed>` object because there's no longer a failure within `cib_process_create() -> add_cib_object()` like there was before.
> 
> Side note, `cibadmin` doesn't support `--output-as=xml` yet. This `<failed>` object only appears as undocumented, schema-less XML in text output.

I believe it's also returned to the API callback (not that anything external is known to use the API). One question is why `update_results()` is called only in `cib_process_create()` and not more broadly.

Reply:
> I believe it's also returned to the API callback 

Yeah looks that way. I'd like to drop `<failed>` and `update_results()` altogether whenever that's allowed. It's used *only* with CIB create operatoins, its format isn't documented or codified in a schema, and it's mostly redundant.

> One question is why `update_results()` is called only in `cib_process_create()` and not more broadly.

It was created specifically for `cib_process_create()` in aa5e6743. Maybe it was planned to be used more extensively later and never was (or maybe it was, and those uses have since disappeared). Or maybe ACLs were implemented differently back then, so that we had to check for denials after adding attributes, but that seems less likely.

Reply:
> > I believe it's also returned to the API callback
> 
> Yeah looks that way. I'd like to drop `<failed>` and `update_results()` altogether whenever that's allowed. It's used _only_ with CIB create operatoins, its format isn't documented or codified in a schema, and it's mostly redundant.

Yeah that should be fine to deprecate for 2.1.8 and drop in 3.0.0. A deprecation message in the API doxygen should be sufficient.

> > One question is why `update_results()` is called only in `cib_process_create()` and not more broadly.
> 
> It was created specifically for `cib_process_create()` in [aa5e674](https://github.com/ClusterLabs/pacemaker/commit/aa5e6743f198aa90cb30a81e5b3a73593bdf3ab8). Maybe it was planned to be used more extensively later and never was (or maybe it was, and those uses have since disappeared). Or maybe ACLs were implemented differently back then, so that we had to check for denials after adding attributes, but that seems less likely.

Reply:
Lol I just noticed https://github.com/ClusterLabs/pacemaker/commit/aa5e6743f198aa90cb30a81e5b3a73593bdf3ab8#diff-43fdbe47b2127355408c6fbf9c07d8b8f2ab141eead9141f900ebf202084658cR1002

"remove this function" for `update_results()`

---

> It was created specifically for `cib_process_create()` in [aa5e674](https://github.com/ClusterLabs/pacemaker/commit/aa5e6743f198aa90cb30a81e5b3a73593bdf3ab8)

This was incorrect. Commit aa5e674 made a duplicate `update_cib_object()` function in a new file (cib_ops.c) without removing the old `update_cib_object()` in `primatives.c` (*sic*).

The original one goes all the way back to 1f442e5d0, where it *was* used in more places.

Mystery solved.

Reply:
> > do you know why it happened?

> been trying to figure out EXACTLY what differences exist...

Following up with more detail on the differing messages from `pcmk__check_acl()`:
* Previously, the denial happened when `update_cib_object()` -> `copy_in_properties()` -> `crm_xml_add()` called `pcmk__check_acl()`. We made it past the creation of the `<primitive>` element, and that was allowed during `pcmk__apply_creation_acl()`. However, the first attempt to set an attribute caused `crm_xml_add()` to error, which in turn caused `update_cib_object()` and `add_cib_object()` to error, leaving an empty `<primitive>` element.
* Now, we don't set any attributes explicitly; we copy the whole subtree. So we get our first error message from `pcmk__check_acl()` later, and it reflects the fact that `<primitive id="dummy">` was copied with its attributes. `cib_perform_op()` -> `xml_acl_disable()` -> `pcmk__apply_creation_acl()` (recursive) -> `pcmk__check_acl()`.